### PR TITLE
changefeed (ticdc): fix owner stuck when closing a changefeed

### LINKED ISSUE #6859: redo log changefeed stuck after network chaos injection
### What did you do?

1. Start redo log changefeed, store redo log to testing minio server
2. Run sysbench workload for 5 miniutes
3. Run "cdc redo apply" to downstream. (I didn't stop the changefeed before redo apply)
4. Inject network chaos between tikv <-> minio, and cdc <-> minio for 10 minio, then restore the chaos injection.
5. About 2 hours later, I use "cdc cli changefeed list" to check the changefeed status

### What did you expect to see?

Changefeed status should be normal

### What did you see instead?

"cdc cli changefeed list" command stucks.

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
| Release Version: v6.2.0-alpha
Edition: Community
Git Commit Hash: 8b5b724d8a932239303a1d0ba547323eb0e5161b
Git Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time: 2022-08-19 11:28:57
GoVersion: go1.18.5
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
```

Upstream TiKV version (execute `tikv-server --version`):

```console
/ # /tikv-server -V
TiKV
Release Version:   6.2.0-alpha
Edition:           Community
Git Commit Hash:   58fa80e0de0d43d473dc456081a4d2b08939e0aa
Git Commit Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time:    2022-08-19 11:03:24
Rust Version:      rustc 1.64.0-nightly (0f4bcadb4 2022-07-30)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-en
gine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
```

TiCDC version (execute `cdc version`):

```console
bash-5.1# /cdc version
Release Version: v6.2.0-alpha
Git Commit Hash: 750f0e4ddf97bb4d9feba66799de93113af04405
Git Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time: 2022-08-22 11:03:29
Go Version: go version go1.18.5 linux/amd64
Failpoint Build: false
```

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #6859 

### What is changed and how it works?
Use `c.isReleased` instead of `c.initialized` to determine whether we should release resources of a changefeed, otherwise if an error occurs in the middle of `c.initialize()`,  the `c.initialized` will be `false`, so the resources of a changefeed would not be release expectantly, which leads to some goroutines leak and can not be cancel by `c.cancel()`.

#### Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)
1. make build-failpoint
2. start cdc server 
3. create a changefeed 
4. error occur when changefeed initializing, and this changefeed will be closed immediately
image
5. `cli changefeed [ pasue ,resume ,remove ]` this changefeed, all commands work well which indicate the owner is working normally.

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/owner/changefeed.go [hunk 1]
```diff
@@ -87,6 +87,8 @@ type changefeed struct {
 	initialized bool
 	// isRemoved is true if the changefeed is removed
 	isRemoved bool
+	// isReleased is true if the changefeed's resource is released
+	isReleased bool
```
Comment (3AceShowHand) [inline cdc/owner/changefeed.go:91]: `initialized` / `isRemoved` / `isReleased`.

Do we really need so many variables?

Comment (asddongmen) [reply cdc/owner/changefeed.go:91]: or we can add a status field for changefeed

File: cdc/owner/changefeed.go [hunk 2]
```diff
@@ -524,37 +537,48 @@ LOOP:
 	return nil
 }
 
+// releaseResources is idempotent.
 func (c *changefeed) releaseResources(ctx cdcContext.Context) {
```
Comment (3AceShowHand) [inline cdc/owner/changefeed.go:541]: can we remove `isReleased`, and still keep `releaseResources` idempotent ?

File: cdc/owner/changefeed.go [hunk 3]
```diff
@@ -524,37 +537,48 @@ LOOP:
...
-	c.wg.Wait()
-	c.scheduler.Close(ctx)
-	c.scheduler = nil
-	c.barriers = nil
 
 	c.cleanupMetrics()
+	c.schema = nil
+	c.barriers = nil
 	c.initialized = false
+	c.isReleased = true
```
Comment (3AceShowHand) [inline cdc/owner/changefeed.go:581]: it looks that `initialized = !isReleased`, so can we only keep one?

Comment (asddongmen) [reply cdc/owner/changefeed.go:581]: the have sligthly different, `isReleased` is assigned to `false` [here](https://github.com/pingcap/tiflow/pull/6882/files#diff-953746562776cf8aa41cd4a971df8c546649d2721ae787607aed2069f3b44eb0R387),  which indicated that we need to release this changefeed if an error occurs in `initialize()`.

**Review by zhaoxinyu — APPROVED** [review]

**Review by nongfushanquan — APPROVED** [review]

File: cdc/owner/ddl_sink.go [hunk 4]
```diff
@@ -338,9 +338,10 @@ func (s *ddlSinkImpl) emitSyncPoint(ctx cdcContext.Context, checkpointTs uint64)
 
 func (s *ddlSinkImpl) close(ctx context.Context) (err error) {
 	s.cancel()
+	// they will both be nil if changefeed return an error in initializing
 	if s.sinkV1 != nil {
 		err = s.sinkV1.Close(ctx)
-	} else {
+	} else if s.sinkV2 != nil {
```
Comment (3AceShowHand) [inline cdc/owner/ddl_sink.go:344]: this may be nil in the test.

**Pull request merged**